mono/metadata/handle.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * handle.h: Handle to object in native code
+ *
+ * Authors:
+ *  - Ludovic Henry <ludovic@xamarin.com>
+ *
+ * Copyright 2015 Xamarin, Inc. (www.xamarin.com)
+ */
+
+#ifndef __MONO_HANDLE_H__
+#define __MONO_HANDLE_H__
+
+#include <config.h>
+#include <glib.h>
+
+#include "object.h"
+#include "class.h"
+#include "class-internals.h"
+#include "threads-types.h"
+
+#include "mono/utils/mono-threads-coop.h"
+
+G_BEGIN_DECLS
+
+typedef struct _MonoHandleStorage MonoHandleStorage;
+typedef MonoHandleStorage* MonoHandle;
+
+/*
+ * DO NOT ACCESS DIRECTLY
+ * USE mono_handle_obj BELOW TO ACCESS OBJ
+ *
+ * The field obj is not private as there is no way to do that
+ * in C, but using a C++ template would simplify that a lot
+ */
+struct _MonoHandleStorage {
+	MonoObject *obj;
+};
+
+#ifndef CHECKED_BUILD
+
+#define mono_handle_obj(handle) ((handle)->obj)
+
+#define mono_handle_assign(handle,rawptr) do { (handle)->obj = (rawptr); } while(0)
+
+#else
+
+static inline void
+mono_handle_check_in_critical_section (const gchar* file, const gint lineno)
+{
+	MonoThreadInfo *info = (MonoThreadInfo*) mono_thread_info_current_unchecked ();
+	if (info && !info->inside_critical_region)
+		g_error ("Assertion at %s:%d: mono_handle_check_in_critical_section failed", file, lineno);
+}
+
+#define mono_handle_obj(handle) (mono_handle_check_in_critical_section (__FILE__, __LINE__), (handle)->obj)
+
+#define mono_handle_assign(handle,rawptr) do { mono_handle_check_in_critical_section (__FILE__, __LINE__); (handle)->obj = (rawptr); } while (0)
+
+#endif
+
+static inline MonoClass*
+mono_handle_class (MonoHandle handle)
+{
+	return handle->obj->vtable->klass;
+}
+
+static inline MonoDomain*
+mono_handle_domain (MonoHandle handle)
+{
+	return handle->obj->vtable->domain;
+}
+
+#define MONO_HANDLE_TYPE_DECL(type)      typedef struct { type *obj; } type ## HandleStorage ; \
+	typedef type ## HandleStorage * type ## Handle
+#define MONO_HANDLE_TYPE(type)           type ## Handle
+#define MONO_HANDLE_NEW(type,obj)        ((type ## Handle) mono_handle_new ((MonoObject*) (obj)))
```

Comment:
where is the `mono_handle_new` function defined? We cannot simply replace it by `mono_handle_arena_new` as that would mean we need to define a new arena in every function that might create a handle. Otherwise we would have to pass it as a parameter to the function, but that would expose it to the user, and we don't necessarily want to do that.